lib/temporal.rb

```diff
@@ -81,6 +83,58 @@ def signal_workflow(workflow, signal, workflow_id, run_id, input = nil)
       )
     end
 
+    # Long polls for a workflow to be completed and returns whatever the execute function
+    # returned.
+    # run_id of nil: await the entire workflow completion.  This can span multiple runs
+    # in the case where the workflow uses continue as new.
+    def await_workflow_result(workflow, workflow_id:, run_id: nil, **args)
+      options = args.delete(:options) || {}
+      execution_options = ExecutionOptions.new(workflow, options)
+
+      current_run_id = run_id
+      loop do
+        history_response = client.get_workflow_execution_history(
+          namespace: execution_options.namespace,
+          workflow_id: workflow_id,
+          run_id: current_run_id,
+          wait_for_new_event: true,
+          event_type: :close
+        )
+        event = history_response['history']['events'].first
+        case event.event_type
+        when :EVENT_TYPE_WORKFLOW_EXECUTION_COMPLETED
+          payloads = event['workflow_execution_completed_event_attributes'].result
+          return nil if !payloads # happens when the workflow itself returns nil
+          return Temporal::Client::Converter::Payload::JSON.new.from_payload(payloads['payloads'].first)
+        when :EVENT_TYPE_WORKFLOW_EXECUTION_TIMED_OUT
+          raise Temporal::WorkflowTimedOut
+        when :EVENT_TYPE_WORKFLOW_EXECUTION_TERMINATED
+          raise Temporal::WorkflowTerminated
+        when :EVENT_TYPE_WORKFLOW_EXECUTION_CANCELED
+          raise Temporal::WorkflowCanceled
+        when :EVENT_TYPE_WORKFLOW_EXECUTION_FAILED
+          # failure_info: Temporal::Api::Failure::V1::Failure
+          failure_info = event['workflow_execution_failed_event_attributes']['failure']
+          raise Temporal::WorkflowFailed.new(
+            failure_info['message'],
```

Comment:
We most likely need the actual error class of the failure here for continuity. Message will not contain that